### ISSUE TITLE
wandb won't automatically do `git commit`

### ISSUE THREAD

**Issue by aiyolo** [wandb/client#1135]
Every time when I run an experiment, I wish wandb will automatically do the commit work for me, but it isn't like that, I wonder what's the purpose for this design?

**Comment by vanpelt** [comment]
Hey @aiyolo can you clarify what you mean be "commit work".  Do you want wandb to automatically run `git commit` for you?  Would you want it to commit directly to the current branch?  What would you want the commit message to be?

We do automatically save a patch file of any uncommitted changes so you can get back to the exact state that your git repo was in for any given run by running `wandb restore username/project/run_id`.

**Comment by wiseyoungbuck** [comment]
As of today, I am receiving the error: when I attempt to log_and artifact.
`wandb: Wandb version 0.9.2 is available!  To upgrade, please run:
wandb:  $ pip install wandb --upgrade
400 response executing GraphQL.
{"errors":[{"message":"cannot use artifact QXJ0aWZhY3Q6NDc2Nw== as it has not committed","path":["useArtifact"]}],"data":{"useArtifact":null}}
wandb: ERROR Error while calling W&B API: cannot use artifact QXJ0aWZhY3Q6NDc2Nw== as it has not committed (<Response [400]>)
`
This was working yesterday. I am using client version 0.9.1

**Comment by annirudh** [comment]
> As of today, I am receiving the error: when I attempt to log_and artifact.
> `wandb: Wandb version 0.9.2 is available! To upgrade, please run: wandb: $ pip install wandb --upgrade 400 response executing GraphQL. {"errors":[{"message":"cannot use artifact QXJ0aWZhY3Q6NDc2Nw== as it has not committed","path":["useArtifact"]}],"data":{"useArtifact":null}} wandb: ERROR Error while calling W&B API: cannot use artifact QXJ0aWZhY3Q6NDc2Nw== as it has not committed (<Response [400]>) `
> This was working yesterday. I am using client version 0.9.1

@wiseyoungbuck Would you mind sharing at a high level what pattern you're using to log an artifact? Our database indicates that `QXJ0aWZhY3Q6NDc2Nw==` was never committed, which can happen if a file failed to upload. If you try reuploading the artifact, subsequent calls to `use_artifact` should work.

I just ran some tests in 0.9.1 using `log_artifact` and `use_artifact` and both seemed to work as expected. I want to make sure I'm running the same steps as you are to reproduce the issue.

**Comment by aiyolo** [comment]
> Hey @aiyolo can you clarify what you mean be "commit work". Do you want wandb to automatically run `git commit` for you? Would you want it to commit directly to the current branch? What would you want the commit message to be?
> 
> We do automatically save a patch file of any uncommitted changes so you can get back to the exact state that your git repo was in for any given run by running `wandb restore username/project/run_id`.

yes, I mean 'git commit', I wish wandb can commit directly to the current branch or create a new branch, the commit message would be wandb.name or something related to the current experiment.

**Comment by aiyolo** [comment]
To the best of my knowledge, the patch file only contains the changes to the current executive file, e.g. `train.py` , so if we change another file 'lib.py',  the changes of 'lib.py' won't be recorded, so `wandb restore` will never restore the old version of 'lib.py'

**Comment by vanpelt** [comment]
Hey @aiyolo the patch file is generated by running `git diff` so it should capture all files currently tracked by git in your repo.

**Comment by EricCousineau-TRI** [comment]
I've also found myself running training, uploading to wandb, without committing.
The patch file sounds great, but I still may have some info that I want to capture in my commit (beyond just a patch) that I'd like to force myself to write down.

One potential workaround (that I may put in my own code) is to add an `argparse` command like `source_tree_mode={strict,devel}`, where `strict` will just error out if there are unstaged / uncommitted changes in the tree.

For git, here's code that I've used that fails fast in the event of a dirty source tree:
```py
status_text = subprocess.check_output(
    "git status --porcelain", shell=True, encoding="utf8").strip()
assert len(status_text) == 0, (
    "This script will not proceed until your worktree is completely " +
    "clean (unstaged and staged files).")
```
